### ISSUE TITLE
Teuchos: Teuchos::as fails to compile with -Werror=double-promotion

### ISSUE THREAD

**Issue by tjfulle** [trilinos/Trilinos#3676]
When compiling with `-Werror=double-promotion`, the following compile time error occurs:

```
            [with OutputRealType = float; InputRealType = double]’:
     312    <path>/include/Teuchos_as.hpp:648:47:   required from here
  >> 313    <path>/include/Teuchos_as.hpp:485:11: error: implicit conversion from ‘float’ to ‘double’ to match other operand of binary expression [-Werror=double-promotion]
     314             x < minVal || x > maxVal,
```
@trilinos/teuchos 

## Expectations
Teuchos compiles with `-Werror=double-promotion`

## Possible Solution
```
diff --git a/packages/teuchos/core/src/Teuchos_as.hpp b/packages/teuchos/core/src/Teuchos_as.hpp
index 3a3798809d..f957a6b735 100644
--- a/packages/teuchos/core/src/Teuchos_as.hpp
+++ b/packages/teuchos/core/src/Teuchos_as.hpp
@@ -482,7 +482,7 @@ namespace { // anonymous
       // let it pass through, per the rules for propagation of silent
       // NaN.  (Signaling NaN will signal, but that's OK.)
       TEUCHOS_TEST_FOR_EXCEPTION(
-        x < minVal || x > maxVal,
+        x < as<InputRealType>(minVal) || x > as<InputRealType>(maxVal),
         std::range_error,
         "realToReal<" << TypeNameTraits<OutputRealType>::name () << ", "
         << TypeNameTraits<InputRealType>::name () << ">: "
```

**Comment by tjfulle** [comment]
@bartlettroscoe

**Comment by bartlettroscoe** [comment]
@tjfulle, what compiler options generated this?  You can find this out by building with `make VERBOSE=1 <object-file-target>`.

**Comment by tjfulle** [comment]
@bartlettroscoe: generated with gcc 7.2.0/openmpi 3.1.2.

I'm not exactly sure the compiler options.  It occurred while installing @ibaned Omega_h library with spack.  I put in a patch to get by the compiler error and rebuilt.  Spack wipes the build directory when done.  I could reproduce it if @ibaned doesn't know the compiler flags

**Comment by ibaned** [comment]
I'm not sure if the rest of the options matter, it seems pretty clear this code is comparing a `double` with a `float` which would trigger `-Wdouble-promotion`.

**Comment by tjfulle** [comment]
Fixed n #3676

**Comment by mhoemmen** [comment]
Fixed by https://github.com/trilinos/Trilinos/pull/3677 -- thanks!
